p11-kit/rpc-message.c

```diff
@@ -767,3 +755,486 @@ p11_rpc_buffer_get_byte_array (p11_buffer *buf,
+	CK_ULONG ulong_value;
+
+	memcpy (&ulong_value, value, value_length);
+	if (ulong_value > 0xFFFFFFFFFFFFFFFF)
+		return 0;
+
+	if (buffer)
+		p11_rpc_buffer_add_uint64 (buffer, ulong_value);
+
+	return 8;
+}
+
+static size_t
+p11_rpc_buffer_add_attribute_value_attribute_array (p11_buffer *buffer,
+						    const void *value,
+						    size_t value_length)
+{
+	const CK_ATTRIBUTE *attrs = value;
+	size_t count = value_length / sizeof (CK_ATTRIBUTE);
+	size_t i, total = 0;
+
+	/* Write the number of items */
+	p11_rpc_buffer_add_uint32 (buffer, count);
+
+	/* Actually write the attributes.  */
+	for (i = 0; i < count; i++)
+		total += p11_rpc_buffer_add_attribute (buffer, &(attrs[i]));
+
+	return 4 + total;
+}
+
+static size_t
+p11_rpc_buffer_add_attribute_value_mechanism_type_array (p11_buffer *buffer,
+							 const void *value,
+							 size_t value_length)
+{
+	const CK_MECHANISM_TYPE *mechs = value;
+	size_t count = value_length / sizeof (CK_MECHANISM_TYPE);
+	size_t i;
+
+	/* Write the number of items */
+	p11_rpc_buffer_add_uint32 (buffer, count);
+
+	for (i = 0; i < count; i++)
+		p11_rpc_buffer_add_uint64 (buffer, mechs[i]);
+
+	return 8 * count;
+}
+
+static size_t
+p11_rpc_buffer_add_attribute_value_date (p11_buffer *buffer,
+					 const void *value,
+					 size_t value_length)
+{
+	CK_DATE date_value;
+
+	memcpy (&date_value, value, value_length);
+
+	if (buffer) {
+		p11_rpc_buffer_add_byte_array (buffer, date_value.year, 4);
+		p11_rpc_buffer_add_byte_array (buffer, date_value.month, 2);
+		p11_rpc_buffer_add_byte_array (buffer, date_value.day, 2);
+	}
+
+	return 8;
+}
+
+static size_t
+p11_rpc_buffer_add_attribute_value_byte_array (p11_buffer *buffer,
+					       const void *value,
+					       size_t value_length)
+{
+	if (buffer)
+		p11_rpc_buffer_add_byte_array (buffer, value, value_length);
+
+	return 4 + value_length;
+}
+
+typedef size_t (*p11_rpc_attribute_value_encoder) (p11_buffer *, const void *, size_t);
+
+static size_t
+p11_rpc_buffer_add_attribute (p11_buffer *buffer, const CK_ATTRIBUTE *attr)
+{
+	unsigned char validity;
+	size_t total = 0;
+	static const p11_rpc_attribute_value_encoder encoders[] = {
+		p11_rpc_buffer_add_attribute_value_byte,
+		p11_rpc_buffer_add_attribute_value_ulong,
+		p11_rpc_buffer_add_attribute_value_attribute_array,
+		p11_rpc_buffer_add_attribute_value_mechanism_type_array,
+		p11_rpc_buffer_add_attribute_value_date,
+		p11_rpc_buffer_add_attribute_value_byte_array
+	};
+	p11_rpc_attribute_value_encoder encoder;
+
+	/* The attribute type */
+	if (buffer)
+		p11_rpc_buffer_add_uint32 (buffer, attr->type);
+	total += 4;
+
+	/* Write out the attribute validity */
+	validity = (((CK_LONG)attr->ulValueLen) == -1) ? 0 : 1;
+	if (buffer)
+		p11_rpc_buffer_add_byte (buffer, validity);
+	total++;
+
+	if (!validity)
+		return total;
+
+	/* The attribute length and value */
+	p11_rpc_buffer_add_uint32 (buffer, attr->ulValueLen);
+	encoder = encoders[map_attribute_type_to_value_type (attr->type)];
```

Comment:
That seems like error prone on future revisions (when new RPC types are added), with a jump to an invalid address confusing debuggers. Maybe separate into an assign part to allow error checking? e.g.
```
if (attr->type > P11_RPC_ATTR_MAX)
  error();

mapval = map_attribute_type_to_value_type (attr->type);
encoder = encoders[mapval]
if (encoder == NULL)
  error();

```